File: nelman-kmcLATTICEMODEL/source/MonteCarlo.cpp
#include "GranCan.hpp"
#include "MonteCarlo.hpp"
#include "Lattice.hpp"
#include "Chain.hpp"
#include "MolBox.hpp"
#include "EnergyMap.hpp"
#include "Stats.hpp"
#include "StatsMethods.hpp"
#include "Cluster.hpp"
#include "ptemp.h"
#include "Moves.hpp"

#include "App.hpp"


#include <string.h>
#include <math.h>
#include <stdio.h>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <fstream>
#include <vector>

#define CORRECTION_0CHAINS


extern Pos local[6];

 

MonteCarlo::MonteCarlo(){
  mainLattice = new Lattice;
  //molbox=new MolBox(mainLattice);  // needed to for Stats::setLattice
  grancan = new GranCan(mainLattice);


  pGlobal =0.1;
  pInsDel =0.01;
  pClusterMove=0.005;

  pChangeStrandCoil=0.5;
  
  steps =10000;
  total_swaps=1;


}



int MonteCarlo::getCext(){
  mainLattice->stats.printCout();
  cout<<"CEXT "<<App::fn_in<<" "<<mainLattice->stats.getCext()<<endl;
  return mainLattice->stats.getCext();

}


int MonteCarlo::MC(double beta, int betaId, bool betaChanged){
  
  grancan->totalEmptySteps=0;
  //  cout<< "starting MC "<<endl;
  
  //mainLattice->writePDBMultiChain("lastConfig.pdb");
  mainLattice->setBetaMoves(beta);
  // cout<< "betas mainLattice set "<<endl;
  if(App::eBetaMu>0){
    grancan->molbox->createNewSet(beta,App::num_configs,betaChanged);
    //cout<< "new config  set created"<<endl;
  }
  mainLattice->energyMap->setBetaID(betaId);

  for(long i=0;i<steps;i++){
    //    cout<< i <<endl;
    //mainLattice->stats.printCout();
    if(drand48() < pGlobal){
      Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
      Moves::globalMove(mainLattice,c);
      //cout<<"finished global move"<<endl;
       // mainLattice->checkStats();
    }
    if(drand48() < pInsDel){   
      if(drand48()<0.5){
		grancan->trialInsertChain();
      }else{
		grancan->trialDeleteChain();
      }
    }
    
    if(drand48() < pChangeStrandCoil){
      Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
      Moves::changeStrandCoil(mainLattice,c);
      //cout<<"finished strand coil switch"<<endl;
      //mainLattice->checkStats();
    }else{
      Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
      Moves::shuffleSpin(mainLattice,c);
      //cout<<"finished spin flip"<<endl;
      //mainLattice->checkStats();
    }
    Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
    Moves::localMove(mainLattice,c);
     //cout<<"finished local move"<<endl;
     //mainLattice->checkStats();
  }

  if(App::clusterMoves && pClusterMove < drand48()){
    grancan->clusterMove();
  }


  for(int c=0;c<mainLattice->nChains;c++){
     mainLattice->chains[c]->checkAllFwdBkwd();
  }
  ClusterInfo ci;
  if(pInsDel >0 || App::clustStats){
    Cluster cl;
    ci = cl.checkCluster(mainLattice,beta);
  }
  mainLattice->checkStats();
  if(App::recordMovie){
    mainLattice->writePDBMultiChain(App::moviestream,App::moviestep);
    App::moviestep++;
  } 


  mainLattice->energyMap->writeTableStats(mainLattice->stats,App::tableStep,1.0,ci);
  if(grancan->totalEmptySteps>0){
    Stats emptyStats;
    double weight = grancan->totalEmptySteps/(double)steps;
    ClusterInfo ci_empty = {0,0,0};
    mainLattice->energyMap->writeTableStats(emptyStats,- App::tableStep,weight,ci_empty);
  }
  App::tableStep++;
  //mainLattice->stats.printCout();
  return mainLattice->stats.getEtot(); 
}



void MonteCarlo::getMCStats(int num_procs, int myRank){
  Cluster cl;
  cl.startCounting(mainLattice);
  cl.printCout();

  std::stringstream outPDB;
  outPDB << "outN"<<myRank<<".pdb";
  mainLattice->writePDBMultiChain(outPDB.str());


  std::stringstream outSTATS;
  outSTATS << App::fn_stats << "N"<<myRank <<".tab"; 
  mainLattice->energyMap->printGnuPlotDataWR(outSTATS.str(),num_procs,myRank);
  
  
  std::stringstream outEMAP;
  outEMAP << App::fn_eMap << "N"<<myRank <<".tab"; 
  mainLattice->energyMap->printEMap(outEMAP.str(),num_procs,myRank);

  if(App::hbondStats){
    std::stringstream outHBondMap;
    outHBondMap << App::fn_hbond << "N"<<myRank <<".tab"; 
    mainLattice->energyMap->printHBondMap(outHBondMap.str(),num_procs,myRank);
  }

  mainLattice->stats.printCout();
  cout<<"free chains: "<<grancan->getNumFreeChains()<<endl;

  mainLattice->energyMap->finalizeTableStats();
    
  //printGnuPlotData(fn_stats,num_procs,myRank);
  //printEMap(fn_eMap,num_procs,myRank);
  //transitionRates->print2file("transitions",myRank);
  //sqrg->print2file("sqrg",myRank);
}

File: nelman-kmcLATTICEMODEL/source/MonteCarlo.py
import random
from typing import Optional
from GranCan import GranCan
from Lattice import Lattice
from Moves import globalMove, changeStrandCoil, shuffleSpin, localMove
from Cluster import Cluster
from App import App

class MonteCarlo:
    def __init__(self):
        self.main_lattice = Lattice()
        self.grancan = GranCan(self.main_lattice)
        
        self.p_global = 0.1
        self.p_ins_del = 0.01
        self.p_cluster_move = 0.005
        self.p_change_strand_coil = 0.5
        
        self.steps = 10000
        self.total_swaps = 1
    
    def getCext(self) -> int:
        self.main_lattice.stats.printCout()
        print(f"CEXT {App.fn_in} {self.main_lattice.stats.getCext()}")
        return self.main_lattice.stats.getCext()
    
    def MC(self, beta: float, beta_id: int, beta_changed: bool) -> int:
        self.grancan.total_empty_steps = 0
        self.main_lattice.setBetaMoves(beta)
        
        if App.eBetaMu > 0:
            self.grancan.molbox.createNewSet(beta, App.num_configs, beta_changed)
        
        self.main_lattice.energy_map.setBetaID(beta_id)
        
        for _ in range(self.steps):
            if random.random() < self.p_global:
                c = random.choice(self.main_lattice.chains[:self.main_lattice.nChains])
                globalMove(self.main_lattice, c)
            
            if random.random() < self.p_ins_del:
                if random.random() < 0.5:
                    self.grancan.trialInsertChain()
                else:
                    self.grancan.trialDeleteChain()
            
            if random.random() < self.p_change_strand_coil:
                c = random.choice(self.main_lattice.chains[:self.main_lattice.nChains])
                changeStrandCoil(self.main_lattice, c)
            else:
                c = random.choice(self.main_lattice.chains[:self.main_lattice.nChains])
                shuffleSpin(self.main_lattice, c)
            
            c = random.choice(self.main_lattice.chains[:self.main_lattice.nChains])
            localMove(self.main_lattice, c)
        
        if App.clusterMoves and random.random() < self.p_cluster_move:
            self.grancan.clusterMove()
        
        # Check chain constraints
        for c in self.main_lattice.chains[:self.main_lattice.nChains]:
            c.checkAllFwdBkwd()
        
        ci = None
        if self.p_ins_del > 0 or App.clustStats:
            cl = Cluster()
            ci = cl.checkCluster(self.main_lattice, beta)
        
        self.main_lattice.checkStats()
        
        if App.recordMovie:
            self.main_lattice.writePDBMultiChain(App.moviestream, App.moviestep)
            App.moviestep += 1
        
        self.main_lattice.energy_map.writeTableStats(
            self.main_lattice.stats, App.tableStep, 1.0, ci
        )
        
        if self.grancan.total_empty_steps > 0:
            empty_stats = Stats()
            weight = self.grancan.total_empty_steps / float(self.steps)
            ci_empty = ClusterInfo(0, 0, 0)
            self.main_lattice.energy_map.writeTableStats(
                empty_stats, -App.tableStep, weight, ci_empty
            )
        
        App.tableStep += 1
        return self.main_lattice.stats.getEtot()
    
    def getMCStats(self, num_procs: int, my_rank: int):
        cl = Cluster()
        cl.startCounting(self.main_lattice)
        cl.printCout()
        
        # Write output files
        out_pdb = f"outN{my_rank}.pdb"
        self.main_lattice.writePDBMultiChain(out_pdb)
        
        out_stats = f"{App.fn_stats}N{my_rank}.tab"
        self.main_lattice.energy_map.printGnuPlotDataWR(out_stats, num_procs, my_rank)
        
        out_emap = f"{App.fn_eMap}N{my_rank}.tab"
        self.main_lattice.energy_map.printEMap(out_emap, num_procs, my_rank)
        
        if App.hbondStats:
            out_hbond = f"{App.fn_hbond}N{my_rank}.tab"
            self.main_lattice.energy_map.printHBondMap(out_hbond, num_procs, my_rank)
        
        self.main_lattice.stats.printCout()
        print(f"free chains: {self.grancan.getNumFreeChains()}")
        
        self.main_lattice.energy_map.finalizeTableStats() 
